### PR TITLE
[#5392] disable "sign transaction" button when offline

### LINKED ISSUE #5392: Sending payment causes TypeError null is not an object when a user has signed while being offline
### Description

[comment]: # (Feature or Bug? i.e Type: Bug)
*Type*: Bug

[comment]: # (Describe the feature you would like, or briefly summarise the bug and what you did, what you expected to happen, and what actually happens. Sections below)
*Summary*: Sending a payment through the chat, when the user has logged into his account while being offline, produces the following error:

```
TypeError: null is not an object (evaluating 't.minus')
    This error is located at:
        in send-transaction
        in reagent193
        in RCTView
        in main
        in root
        in reagent1
        in RCTView
        in RCTView
        in t
```

This error produces the application to freeze, showing a blank screen. It is necessary to close the application in order to continue using it.

Here is a screen capture with the Error dialog shown to the user:

device-2018-08-06-125517
 
Here is a screen capture of the wallet:
device-2018-08-06-125544

Here is a video reproducing the bug:
https://drive.google.com/file/d/1T6hYW85u5KyUngAqYZPTvNMPhmNxWHNa/view?usp=sharing

This issue was reproducible with two different Android devices. Not been able to check on iOS devices.

The issue was reproduced 4 times/4 attempts.

#### Expected behavior
The expected behavior would have been to have a dialog telling the user that he/she can't make the transaction because the app is in offline mode.

#### Actual behavior
The application crashes, shows an error dialog and stays in a blank screen. The user must hit the back button two times to be able to quit the application.

### Reproduction
- Disable Wi-Fi
- Disable Mobile data
- Open Status
- Log into existing account
- Open existing chat with a contact
- Send a payment

### Additional Information
* Status version: 0.9.24 (6291)
* Operating System: Android
* Devices: Huawei ALE-L21 with Android 6.0, Nexus 5 with Android 6.0.1

#### Logs

[comment]: # (if on Android please replicate bug whilst running adb logcat)
```
 NAME: "TypeError"
    MESSAGE: TypeError: null is not an object (evaluating 't.minus')
    
    This error is located at:
        in send-transaction
        in reagent193
        in RCTView
        in main
        in root
        in reagent1
        in RCTView
        in RCTView
        in t
    
    check_sufficient_gas@index.android.bundle:12:4927496
    G__30386_30398@index.android.bundle:12:4928796
    index.android.bundle:12:2486676
    index.android.bundle:12:2486688
    in_context@index.android.bundle:12:2286742
    deref_capture@index.android.bundle:12:2286885
    _run@index.android.bundle:12:2307407
    cljs$core$IDeref$_deref$arity$1@index.android.bundle:12:2312393
    _deref@index.android.bundle:12:153854
    deref@index.android.bundle:12:211756
    index.android.bundle:12:7631063
    index.android.bundle:12:2321509
    wrap_render@index.android.bundle:12:2322190
    do_render@index.android.bundle:12:2323372
    index.android.bundle:12:2323925
    in_context@index.android.bundle:12:2286742
    deref_capture@index.android.bundle:12:2286885
    run_in_reaction@index.android.bundle:12:2315218
    index.android.bundle:12:2323866
    f@index.android.bundle:143:83728
    beginWork@index.android.bundle:143:87321
    o@index.android.bundle:143:128242
    a@index.android.bundle:143:128621
    T@index.android.bundle:143:132129
    y@index.android.bundle:143:131565
    v@index.android.bundle:143:131402
    h@index.android.bundle:143:130681
    c@index.android.bundle:143:129935
    enqueueForceUpdate@index.android.bundle:143:60452
    forceUpdate@index.android.bundle:109:1315
    run_queue@index.android.bundle:12:2282434
    flush_queues@index.android.bundle:12:2283706
    run_queues@index.android.bundle:12:2283853
    index.android.bundle:12:2283282
    index.android.bundle:126:1369
    x@index.android.bundle:126:643
    callTimers@index.android.bundle:126:2788
    value@index.android.bundle:121:3582
    index.android.bundle:121:1067
    value@index.android.bundle:121:3009
    value@index.android.bundle:121:1039
    [native code]
```

<blockquote><div>Google Drive</div><div><strong><a href="https://drive.google.com/file/d/1T6hYW85u5KyUngAqYZPTvNMPhmNxWHNa/view?usp=sharing&usp=embed_facebook">offline_null_currency_error.mp4</a></strong></div></blockquote>

### PR DESCRIPTION
**Pull request by dmitryn**
fixes #5392

### Summary:

Disable "sign transaction" button when offline.

status: ready

### CODE REVIEW

**Review by janherich — APPROVED** [review]

File: src/status_im/ui/screens/wallet/send/views.cljs [hunk 1]
```diff
@@ -106,18 +107,22 @@
...
+                                              (not sign-enabled?)
+                                              (= :offline network-status))
                      :accessibility-label :sign-transaction-button}
       (i18n/label sign-label)
       [vector-icons/icon :icons/forward {:color :white}]]]))
 
 ;; "Sign Transaction >" button
 (defn- sign-transaction-button [amount-error to amount sufficient-funds? sufficient-gas? modal?]
-  (let [sign-enabled? (and (nil? amount-error)
+  (let [online? (= :online @(re-frame/subscribe [:network-status]))
```
Comment (flexsurfer) [inline src/status_im/ui/screens/wallet/send/views.cljs]: this looks weird, can we pass online? from outside?

**Review by flexsurfer — CHANGES_REQUESTED** [review]

File: src/status_im/ui/screens/wallet/send/views.cljs (hunk 1 above)
Comment (dmitryn) [reply src/status_im/ui/screens/wallet/send/views.cljs]: @flexsurfer changed, please take a look

**Review by flexsurfer — APPROVED** [review]

**Comment by asemiankevich** [comment]
@dmitryn `transaction underpriced` is received in the following case:

- open status
- faucet some funds
- open 1x1 chat
- go offline
- tap send ETH 0 and go to transaction screen. Note that `sign transaction` button is disabled
- go online (turn off airplane mode)
- check that sign button is available
- tap sign transaction

as a result, user sees `transaction underpriced` pop-up while expecting the transaction is sent. 

Seems gas limit / price are not fetched correctly (i do see values in UI but seems they are not taken into account)

**Comment by dmitryn** [comment]
@asemiankevich i've hidden submit button for /send /receive command when offline, please take a look.

**Comment by lukaszfryc** [comment]
@dmitryn I could not reproduce Nastya's issue. But, I noticed 2 other things that maybe related to this PR.

I had a poor network connection and couple of times I experienced:
1. `t.minus` error while sending ETH in 1-1 chat



I could not found TF session for it.

2. 0 gas price even when my network connection returned back:



[TF](https://app.testfairy.com/projects/4803590-status/builds/8594906/sessions/4401667113/?accessToken=OlCsh8hjn3k/L08u9UK1MCfHtMA) for 0 gas.


Do you think those issues are related?

**Comment by lukaszfryc** [comment]
@dmitryn I actually think those issues are not related to your change. I see them in other PRs as well.

**Comment by lukaszfryc** [comment]
@dmitryn but `TypeError: null is not an object (evaluating 't.minus')` can still be replicated by following these steps:

- Create account, add a contact, switch to ropsten and request ETH
- Kill the app 
- Disable Wi-Fi and Mobile data
- Open Status
- Log into the existing account
- Go to Wallet and send transaction
- Select the contact
- Try to type some ETH amount 

-> The error appears. I think it maybe related to the fact that wallet and send transactions do not have amount, gas limit, etc. fetched because the device was offline. Anyway, this error should not appear.

**Comment by dmitryn** [comment]
@lukaszfryc i've disabled amount input when offline to fix issue you found. Please take a look.

**Comment by lukaszfryc** [comment]
@dmitryn 

1) When I followed steps from my recent comment https://github.com/status-im/status-react/pull/5817#issuecomment-422021405, opened wallet and went through wallet onboarding I got this error:




2) Another thing, the amount is disabled in send transaction screen as you mentioned. But, I strongly believe that we should inform user why we disabled it. How about having a gray bar as we have in Wallet saying "Can't send transaction in offline mode":

**Comment by dmitryn** [comment]
@lukaszfryc please take a look.
1. fixed
2. good idea, added snackbar with that message when offline.

**Comment by lukaszfryc** [comment]
> good idea, added snackbar with that message when offline.

1) The snackbar looks is really nice! Can you add it to "Sign Message" screen as well? Copy can say "Can't sign message in offline mode"

2) I got `null is not an object (evaluating 't.minus')` once again after I installed the app, switched to ropsten requested eth in test dapp. Here is [TF, see from 4:30](https://app.testfairy.com/projects/4803622-status/builds/8596958/sessions/4401762042/?accessToken=bUHwcVnuR3oLKLu1CKjRdxRyChQ). I could not replicate it afterwards :/ 

```
09-18 15:48:49.567 15566 15702 E ReactNativeJS: TypeError: null is not an object (evaluating 't.minus')
09-18 15:48:49.567 15566 15702 E ReactNativeJS:
09-18 15:48:49.567 15566 15702 E ReactNativeJS: This error is located at:
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in send-transaction-modal
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in reagent115
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in RCTView
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in RCTView
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in reagent14
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in RCTView
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in main-screen-modal-view
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in reagent110
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in RCTView
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in RCTModalHostView
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in t
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in RCTView
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in main-modal
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in reagent3
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in RCTView
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in main
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in root
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in reagent1
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in RCTView
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in RCTView
09-18 15:48:49.567 15566 15702 E ReactNativeJS:     in t
```

**Comment by dmitryn** [comment]
@lukaszfryc snackbar added and error fixed (hopefully). Please take a look.

**Comment by lukaszfryc** [comment]
@dmitryn 

> I got null is not an object (evaluating 't.minus') once again after I installed the app, switched to ropsten requested eth in test dapp. Here is TF, see from 4:30. I could not replicate it afterwards :/

I could not reproduce it so I think it's fixed now 👍 

----

But, there are still to do:

> The snackbar looks is really nice! Can you add it to "Sign Message" screen as well? Copy can say "Can't sign message in offline mode"

1) Looks like there is no copy for sign message bar. 



2) On iPhone, the "Amount" field is not clickable after app goes back online. To reproduce go to sign transaction, go offline and then back online. The amount field should be clickable now but it is not.

**Comment by dmitryn** [comment]
@lukaszfryc 

1. added translation
2. There seems a bug when with returning from background on IOS – app state doesn't change back to "online" (i've tested it in develop and have same issue, i even see "Can't send when offline" snackbar when i return back online. It'll be helpful if you could reproduce it on real ios device in develop branch). 
So i switched to not disabling amount field when offline, this doesn't give my any errors and field is still editable when online.

Cannot make build at the moment (jenkins issue), will ping you when new build will be available.

**Comment by dmitryn** [comment]
@lukaszfryc new build is ready, please check.

**Comment by lukaszfryc** [comment]
@dmitryn I'll come back to it later today once [release blockers](https://github.com/status-im/status-react/labels/release) are fixed.

**Comment by lukaszfryc** [comment]
@dmitryn [latest bugs](https://github.com/status-im/status-react/pull/5817#issuecomment-422788765) are fixed but there is a new one. Hopefully the last one as I could not find anything else :) 

On Android (I'm testing on 8.0), when you **device press back button** in transaction fee screen, there is an error:





This only happen when back button is pressed. Works fine for "X" or "Done" buttons. 

I checked latest nightly (`ce741d7`) and all works fine there so I assume it's related to our PR.

**Comment by dmitryn** [comment]
@lukaszfryc i've rebased branch to develop and this error is gone. Waiting for Jenkins build.

**Comment by dmitryn** [comment]
@lukaszfryc build is ready, please check.

**Comment by lukaszfryc** [comment]
I’m on it.

**Comment by lukaszfryc** [comment]
@dmitryn I did not notice any regressions or new issues. Feel free to merge 💯

**Comment by lukaszfryc** [comment]
@dmitryn please cherry pick it to `release/0.9.28` as well.

**Comment by dmitryn** [comment]
@lukaszfryc sure, cherry-picked to release/0.9.28. Thanks for testing!

**Pull request merged**